controllers/twist_controller/src/twist_controller.cpp

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "twist_controller.hpp"
+
+namespace kuka_controllers
+{
+controller_interface::CallbackReturn TwistController::on_init()
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration TwistController::
+command_interface_configuration()
+const
+{
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+  // config.names.emplace_back("twist commands");
+  config.names.emplace_back("twist/linear/x");
+  config.names.emplace_back("twist/linear/y");
+  config.names.emplace_back("twist/linear/z");
+  config.names.emplace_back("twist/angular/x");
+  config.names.emplace_back("twist/angular/y");
+  config.names.emplace_back("twist/angular/z");
+  return config;
+}
+
+controller_interface::InterfaceConfiguration TwistController::state_interface_configuration()
+const
+{
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+  config.names.emplace_back("twist/linear/x");
+  config.names.emplace_back("twist/linear/y");
+  config.names.emplace_back("twist/linear/z");
+  config.names.emplace_back("twist/angular/x");
+  config.names.emplace_back("twist/angular/y");
+  config.names.emplace_back("twist/angular/z");
+  config.names.emplace_back("Pose/position/x");
+  config.names.emplace_back("Pose/position/y");
+  config.names.emplace_back("Pose/position/z");
+  config.names.emplace_back("Pose/orientation/x");
+  config.names.emplace_back("Pose/orientation/y");
+  config.names.emplace_back("Pose/orientation/z");
+  config.names.emplace_back("Pose/orientation/w");
+
+  return config;
+}
+
+controller_interface::CallbackReturn
+TwistController::on_configure(const rclcpp_lifecycle::State &)
+{
+  twist_subscriber_ = get_node()->create_subscription<geometry_msgs::msg::Twist>(
+    "twist_controller/twist_cmd", rclcpp::SystemDefaultsQoS(),
+    [this](const geometry_msgs::msg::Twist::SharedPtr msg) {
+        last_command_msg_ .linear.x= msg->linear.x;
+        last_command_msg_ .linear.y= msg->linear.y;
+        last_command_msg_ .linear.z= msg->linear.z;
+        last_command_msg_ .angular.x= msg->angular.x;
+        last_command_msg_ .angular.y= msg->angular.y;
+        last_command_msg_ .angular.z= msg->angular.z;
+    });
+  return controller_interface::CallbackReturn::SUCCESS;
+
+}
+
+controller_interface::CallbackReturn
+TwistController::on_activate(const rclcpp_lifecycle::State &)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn
+TwistController::on_deactivate(const rclcpp_lifecycle::State &)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type TwistController::update(
```

Comment:
this controller has two well-separable tasks:
- forward commands received on a topic to the hwif
- broadcast tf information
I think it would be beneficial to separate these two tasks. (e.g. for robot arms, there is a separate `joint_trajectory_controller` and a `joint_state_broadcaster` that allows exchanging only one of the control and broadcasting logic.)

Another thing to consider is how much can be reused from already existing code. I have looked at the [steering_controllers_library](https://github.com/ros-controls/ros2_controllers/blob/master/steering_controllers_library/doc/userdoc.rst), which seems to convert the `Twist` commands to joint level, so is not appropriate for this task, but you could simply extend the [forward_command_controller](https://github.com/ros-controls/ros2_controllers/blob/master/forward_command_controller/doc/userdoc.rst) for the controlling part and reduce your code considerably.
For the broadcaster part there is no base I know of, so that can remain, however in a separate controller